Flaky test TestEventListener.testPlanningFailure

https://github.com/prestosql/presto/runs/1386057381

```
2020-11-11T16:45:57.6822217Z [ERROR] Tests run: 2082, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 2,162.941 s <<< FAILURE! - in TestSuite
2020-11-11T16:45:57.6823688Z [ERROR] testPlanningFailure(io.prestosql.execution.TestEventListener)  Time elapsed: 0.019 s  <<< FAILURE!
2020-11-11T16:45:57.6825688Z java.lang.IllegalArgumentException: expected one element but was: <io.prestosql.spi.eventlistener.QueryCompletedEvent@7142368e, io.prestosql.spi.eventlistener.QueryCompletedEvent@4b216d17>
2020-11-11T16:45:57.6827721Z 	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:315)
2020-11-11T16:45:57.6829025Z 	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
2020-11-11T16:45:57.6830576Z 	at io.prestosql.execution.TestEventListener.assertFailedQuery(TestEventListener.java:243)
2020-11-11T16:45:57.6832346Z 	at io.prestosql.execution.TestEventListener.assertFailedQuery(TestEventListener.java:235)
2020-11-11T16:45:57.6834519Z 	at io.prestosql.execution.TestEventListener.testPlanningFailure(TestEventListener.java:164)
2020-11-11T16:45:57.6839232Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-11-11T16:45:57.6846082Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-11-11T16:45:57.6854238Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-11-11T16:45:57.6860892Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-11-11T16:45:57.6869367Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-11-11T16:45:57.6876055Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-11-11T16:45:57.6881417Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-11-11T16:45:57.6886531Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-11-11T16:45:57.6892121Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-11-11T16:45:57.6897339Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-11-11T16:45:57.6904472Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-11-11T16:45:57.6912367Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-11-11T16:45:57.6917184Z 	at java.base/java.lang.Thread.run(Thread.java:834)
```

I'm afraid it's still flaky: https://github.com/prestosql/presto/pull/5355/checks?check_run_id=1472968160

```
2020-11-30T10:22:50.1440796Z [ERROR] Tests run: 2086, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 2,166.121 s <<< FAILURE! - in TestSuite
2020-11-30T10:22:50.1461232Z [ERROR] testPlanningFailure(io.prestosql.execution.TestEventListenerBasic)  Time elapsed: 1.039 s  <<< FAILURE!
2020-11-30T10:22:50.1469840Z java.lang.IllegalArgumentException: expected one element but was: <QueryCompletedEvent[queryId='20201130_101733_00008_x6yan', queryState='FAILED'], QueryCompletedEvent[queryId='20201130_101733_00008_x6yan', queryState='FAILED']>
2020-11-30T10:22:50.1477318Z 	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:315)
2020-11-30T10:22:50.1488719Z 	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
2020-11-30T10:22:50.1495324Z 	at io.prestosql.execution.TestEventListenerBasic.assertFailedQuery(TestEventListenerBasic.java:224)
2020-11-30T10:22:50.1501657Z 	at io.prestosql.execution.TestEventListenerBasic.assertFailedQuery(TestEventListenerBasic.java:216)
2020-11-30T10:22:50.1510641Z 	at io.prestosql.execution.TestEventListenerBasic.testPlanningFailure(TestEventListenerBasic.java:145)
2020-11-30T10:22:50.1518916Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-11-30T10:22:50.1542029Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-11-30T10:22:50.1548422Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-11-30T10:22:50.1553249Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-11-30T10:22:50.1559439Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-11-30T10:22:50.1564605Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-11-30T10:22:50.1569327Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-11-30T10:22:50.1574189Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-11-30T10:22:50.1579669Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-11-30T10:22:50.1584511Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-11-30T10:22:50.1590280Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-11-30T10:22:50.1596063Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-11-30T10:22:50.1600972Z 	at java.base/java.lang.Thread.run(Thread.java:834)
2020-11-30T10:22:50.1601609Z 
2020-11-30T10:22:50.1665431Z 2020-11-30T04:22:50.166-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1669718Z 2020-11-30T04:22:50.166-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1674297Z 2020-11-30T04:22:50.167-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1675382Z 2020-11-30T04:22:50.167-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1706367Z 2020-11-30T04:22:50.170-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1707856Z 2020-11-30T04:22:50.170-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1713457Z 2020-11-30T04:22:50.170-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1714671Z 2020-11-30T04:22:50.170-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1718506Z 2020-11-30T04:22:50.171-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1720416Z 2020-11-30T04:22:50.171-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1721967Z 2020-11-30T04:22:50.171-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1725349Z 2020-11-30T04:22:50.171-0600 INFO JVM is shutting down, cleaning up
2020-11-30T10:22:50.1726463Z 2020-11-30T04:22:50.172-0600 INFO Life cycle stopping...
2020-11-30T10:22:50.1728160Z 2020-11-30T04:22:50.172-0600 INFO Life cycle stopping...
2020-11-30T10:22:53.2979565Z [INFO] 
2020-11-30T10:22:53.2980300Z [INFO] Results:
2020-11-30T10:22:53.2980795Z [INFO] 
2020-11-30T10:22:53.2981290Z [ERROR] Failures: 
2020-11-30T10:22:53.3027556Z [ERROR]   TestEventListenerBasic.testPlanningFailure:145->assertFailedQuery:216->assertFailedQuery:224 » IllegalArgument
2020-11-30T10:22:53.3029112Z [INFO] 
2020-11-30T10:22:53.3029651Z [ERROR] Tests run: 2086, Failures: 1, Errors: 0, Skipped: 1
2020-11-30T10:22:53.3030156Z [INFO] 
2020-11-30T10:22:53.3066563Z [INFO] ------------------------------------------------------------------------
2020-11-30T10:22:53.3067476Z [INFO] BUILD FAILURE
2020-11-30T10:22:53.3068461Z [INFO] ------------------------------------------------------------------------
2020-11-30T10:22:53.3080063Z [INFO] Total time:  36:15 min
2020-11-30T10:22:53.3083142Z [INFO] Finished at: 2020-11-30T10:22:53Z
2020-11-30T10:22:53.3084230Z [INFO] ------------------------------------------------------------------------
```

--- a/testing/trino-tests/src/test/java/io/trino/execution/EventsCollector.java
+++ b/testing/trino-tests/src/test/java/io/trino/execution/EventsCollector.java
@@ -18,6 +18,9 @@ import io.trino.spi.eventlistener.QueryCompletedEvent;
 import io.trino.spi.eventlistener.QueryCreatedEvent;
 import io.trino.spi.eventlistener.SplitCompletedEvent;
 
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
@@ -25,13 +28,17 @@ import java.util.function.Predicate;
 
 import static java.util.Objects.requireNonNull;
 
-class EventsCollector
+@ThreadSafe
+final class EventsCollector
 {
-    private EventFilters eventFilters;
+    private final EventFilters eventFilters;
+    @GuardedBy("this")
     private ImmutableList.Builder<QueryCreatedEvent> queryCreatedEvents;
+    @GuardedBy("this")
     private ImmutableList.Builder<QueryCompletedEvent> queryCompletedEvents;
+    @GuardedBy("this")
     private ImmutableList.Builder<SplitCompletedEvent> splitCompletedEvents;
-
+    @GuardedBy("this")
     private CountDownLatch eventsLatch;
 
     public EventsCollector()
@@ -57,12 +64,17 @@ class EventsCollector
     public void waitForEvents(int timeoutSeconds)
             throws InterruptedException
     {
+        CountDownLatch eventsLatch;
+        synchronized (this) {
+            // since the eventsLatch is replaced in the reset method, a lock is required for proper memory visibility
+            eventsLatch = this.eventsLatch;
+        }
         eventsLatch.await(timeoutSeconds, TimeUnit.SECONDS);
     }
 
     public synchronized void addQueryCreated(QueryCreatedEvent event)
     {
-        if (!eventFilters.queryCreatedFilter.test(event)) {
+        if (!eventFilters.getQueryCreatedFilter().test(event)) {
             return;
         }
         queryCreatedEvents.add(event);
@@ -71,7 +83,7 @@ class EventsCollector
 
     public synchronized void addQueryCompleted(QueryCompletedEvent event)
     {
-        if (!eventFilters.queryCompletedFilter.test(event)) {
+        if (!eventFilters.getQueryCompletedFilter().test(event)) {
             return;
         }
         queryCompletedEvents.add(event);
@@ -80,39 +92,58 @@ class EventsCollector
 
     public synchronized void addSplitCompleted(SplitCompletedEvent event)
     {
-        if (!eventFilters.splitCompletedFilter.test(event)) {
+        if (!eventFilters.getSplitCompletedFilter().test(event)) {
             return;
         }
         splitCompletedEvents.add(event);
         eventsLatch.countDown();
     }
 
-    public List<QueryCreatedEvent> getQueryCreatedEvents()
+    public synchronized List<QueryCreatedEvent> getQueryCreatedEvents()
     {
         return queryCreatedEvents.build();
     }
 
-    public List<QueryCompletedEvent> getQueryCompletedEvents()
+    public synchronized List<QueryCompletedEvent> getQueryCompletedEvents()
     {
         return queryCompletedEvents.build();
     }
 
-    public List<SplitCompletedEvent> getSplitCompletedEvents()
+    public synchronized List<SplitCompletedEvent> getSplitCompletedEvents()
     {
         return splitCompletedEvents.build();
     }
 
+    @ThreadSafe
     public static class EventFilters
     {
         private final Predicate<QueryCreatedEvent> queryCreatedFilter;
         private final Predicate<QueryCompletedEvent> queryCompletedFilter;
         private final Predicate<SplitCompletedEvent> splitCompletedFilter;
 
-        private EventFilters(Predicate<QueryCreatedEvent> queryCreatedFilter, Predicate<QueryCompletedEvent> queryCompletedFilter, Predicate<SplitCompletedEvent> splitCompletedFilter)
+        private EventFilters(
+                Predicate<QueryCreatedEvent> queryCreatedFilter,
+                Predicate<QueryCompletedEvent> queryCompletedFilter,
+                Predicate<SplitCompletedEvent> splitCompletedFilter)
         {
             this.queryCreatedFilter = requireNonNull(queryCreatedFilter, "queryCreatedFilter is null");
             this.queryCompletedFilter = requireNonNull(queryCompletedFilter, "queryCompletedFilter is null");
             this.splitCompletedFilter = requireNonNull(splitCompletedFilter, "splitCompletedFilter is null");
+        }
+
+        Predicate<QueryCreatedEvent> getQueryCreatedFilter()
+        {
+            return queryCreatedFilter;
+        }
+
+        Predicate<QueryCompletedEvent> getQueryCompletedFilter()
+        {
+            return queryCompletedFilter;
+        }
+
+        Predicate<SplitCompletedEvent> getSplitCompletedFilter()
+        {
+            return splitCompletedFilter;
         }
 
         public static Builder builder()

--- a/testing/trino-tests/src/test/java/io/trino/execution/TestCompletedEventWarnings.java
+++ b/testing/trino-tests/src/test/java/io/trino/execution/TestCompletedEventWarnings.java
@@ -14,83 +14,85 @@
 package io.trino.execution;
 
 import com.google.common.collect.ImmutableMap;
-import io.trino.Session.SessionBuilder;
+import com.google.common.io.Closer;
+import io.trino.execution.EventsCollector.EventFilters;
 import io.trino.execution.TestEventListenerPlugin.TestingEventListenerPlugin;
 import io.trino.execution.warnings.WarningCollectorConfig;
 import io.trino.spi.TrinoWarning;
 import io.trino.spi.WarningCode;
 import io.trino.testing.DistributedQueryRunner;
-import io.trino.testing.QueryRunner;
 import io.trino.testing.TestingWarningCollector;
 import io.trino.testing.TestingWarningCollectorConfig;
 import org.intellij.lang.annotations.Language;
-import org.testng.annotations.AfterMethod;
-import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
+import java.io.IOException;
+import java.time.Duration;
 import java.util.List;
-import java.util.Map;
 import java.util.Set;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static com.google.common.collect.Iterables.getOnlyElement;
-import static io.trino.testing.TestingSession.testSessionBuilder;
+import static io.trino.SessionTestUtils.TEST_SESSION;
 import static org.testng.Assert.fail;
 
 @Test(singleThreaded = true)
 public class TestCompletedEventWarnings
 {
-    private static final int EXPECTED_EVENTS = 3;
     private static final int TEST_WARNINGS = 5;
-    private QueryRunner queryRunner;
-    private EventsCollector generatedEvents;
 
-    @BeforeMethod
+    private final EventsCollector generatedEvents = new EventsCollector(EventFilters.builder()
+            .setQueryCreatedFilter(event -> false)
+            .setSplitCompletedFilter(event -> false)
+            .build());
+
+    private Closer closer;
+    private EventsAwaitingQueries queries;
+
+    @BeforeClass
     public void setUp()
             throws Exception
     {
-        SessionBuilder sessionBuilder = testSessionBuilder();
-        generatedEvents = new EventsCollector();
-        queryRunner = DistributedQueryRunner.builder(sessionBuilder.build())
+        closer = Closer.create();
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(TEST_SESSION)
                 .setExtraProperties(ImmutableMap.of("testing-warning-collector.preloaded-warnings", String.valueOf(TEST_WARNINGS)))
                 .setNodeCount(1)
                 .build();
+        closer.register(queryRunner);
         queryRunner.installPlugin(new TestingEventListenerPlugin(generatedEvents));
-        generatedEvents.reset(EXPECTED_EVENTS);
+        queries = new EventsAwaitingQueries(generatedEvents, queryRunner, Duration.ofSeconds(1));
     }
 
-    @AfterMethod(alwaysRun = true)
+    @AfterClass(alwaysRun = true)
     public void tearDown()
+            throws IOException
     {
-        queryRunner.close();
-        queryRunner = null;
-        generatedEvents = null;
+        if (closer != null) {
+            closer.close();
+        }
+        closer = null;
     }
 
     @Test
     public void testCompletedEventWarnings()
-            throws InterruptedException
+            throws Exception
     {
         TestingWarningCollectorConfig warningCollectorConfig = new TestingWarningCollectorConfig().setPreloadedWarnings(TEST_WARNINGS);
         TestingWarningCollector testingWarningCollector = new TestingWarningCollector(new WarningCollectorConfig(), warningCollectorConfig);
         assertWarnings(
                 "select 1",
-                ImmutableMap.of(),
                 testingWarningCollector.getWarnings().stream()
                         .map(TrinoWarning::getWarningCode)
                         .collect(toImmutableList()));
     }
 
-    private void assertWarnings(@Language("SQL") String sql, Map<String, String> sessionProperties, List<WarningCode> expectedWarnings)
-            throws InterruptedException
+    private void assertWarnings(@Language("SQL") String sql, List<WarningCode> expectedWarnings)
+            throws Exception
     {
-        // Task concurrency must be 1 otherwise these tests fail due to change in the number of EXPECTED_EVENTS
-        SessionBuilder sessionBuilder = testSessionBuilder()
-                .setSystemProperty("task_concurrency", "1");
-        sessionProperties.forEach(sessionBuilder::setSystemProperty);
-        queryRunner.execute(sessionBuilder.build(), sql);
-        generatedEvents.waitForEvents(10);
+        queries.runQueryAndWaitForEvents(sql, 1, TEST_SESSION);
 
         Set<WarningCode> warnings = getOnlyElement(generatedEvents.getQueryCompletedEvents())
                 .getWarnings()

--- a/testing/trino-tests/src/test/java/io/trino/execution/TestConnectorEventListener.java
+++ b/testing/trino-tests/src/test/java/io/trino/execution/TestConnectorEventListener.java
@@ -14,40 +14,42 @@
 package io.trino.execution;
 
 import com.google.common.collect.ImmutableList;
-import io.trino.Session;
+import com.google.common.io.Closer;
 import io.trino.connector.MockConnectorFactory;
+import io.trino.execution.EventsCollector.EventFilters;
 import io.trino.spi.Plugin;
 import io.trino.spi.connector.ConnectorFactory;
+import io.trino.spi.eventlistener.QueryCompletedEvent;
+import io.trino.spi.eventlistener.QueryCreatedEvent;
 import io.trino.testing.DistributedQueryRunner;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
+import java.io.IOException;
 import java.time.Duration;
+import java.util.List;
 
-import static io.trino.testing.TestingSession.testSessionBuilder;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.SessionTestUtils.TEST_SESSION;
 import static org.assertj.core.api.Assertions.assertThat;
 
 @Test(singleThreaded = true)
 public class TestConnectorEventListener
 {
-    private final EventsCollector generatedEvents = new EventsCollector();
+    private final EventsCollector generatedEvents = new EventsCollector(EventFilters.builder()
+            .setSplitCompletedFilter(event -> false)
+            .build());
 
-    private DistributedQueryRunner queryRunner;
-    private Session session;
+    private Closer closer;
     private EventsAwaitingQueries queries;
 
     @BeforeClass
     public void setUp()
             throws Exception
     {
-        session = testSessionBuilder()
-                .setSystemProperty("task_concurrency", "1")
-                .setCatalog("tpch")
-                .setSchema("tiny")
-                .setClientInfo("{\"clientVersion\":\"testVersion\"}")
-                .build();
-        queryRunner = DistributedQueryRunner.builder(session).setNodeCount(1).build();
+        closer = Closer.create();
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(TEST_SESSION).setNodeCount(1).build();
         queryRunner.installPlugin(new Plugin()
         {
             @Override
@@ -58,26 +60,39 @@ public class TestConnectorEventListener
                         .build());
             }
         });
+        closer.register(queryRunner);
         queryRunner.createCatalog("mock-catalog", "mock");
         queries = new EventsAwaitingQueries(generatedEvents, queryRunner, Duration.ofSeconds(1));
     }
 
     @AfterClass(alwaysRun = true)
     public void tearDown()
+            throws IOException
     {
-        queryRunner.close();
-        queryRunner = null;
+        if (closer != null) {
+            closer.close();
+        }
+        closer = null;
     }
 
     @Test
     public void testConnectorEventHandlerReceivingEvents()
             throws Exception
     {
-        queries.runQueryAndWaitForEvents("SELECT 1", 3, session);
+        queries.runQueryAndWaitForEvents("SELECT 1", 2, TEST_SESSION);
 
-        assertThat(generatedEvents.getQueryCreatedEvents())
-                .size().isEqualTo(1);
-        assertThat(generatedEvents.getQueryCompletedEvents())
-                .size().isEqualTo(1);
+        List<QueryCreatedEvent> queryCreatedEvents = generatedEvents.getQueryCreatedEvents();
+        List<QueryCompletedEvent> queryCompletedEvents = generatedEvents.getQueryCompletedEvents();
+        List<Object> allEvents = ImmutableList.builder()
+                .addAll(queryCreatedEvents)
+                .addAll(queryCompletedEvents)
+                .build();
+        List<String> eventTypes = allEvents.stream().map(event -> event.getClass().getSimpleName()).sorted().collect(toImmutableList());
+        assertThat(allEvents)
+                .size().withFailMessage(() -> "got events: " + eventTypes).isEqualTo(2);
+        assertThat(queryCreatedEvents)
+                .size().withFailMessage(() -> "got events: " + eventTypes).isEqualTo(1);
+        assertThat(queryCompletedEvents)
+                .size().withFailMessage(() -> "got events: " + eventTypes).isEqualTo(1);
     }
 }

